monitoring/src/prometheus.ts

```diff
@@ -0,0 +1,135 @@
+import * as k8s from '@pulumi/kubernetes'
+import * as pulumi from '@pulumi/pulumi'
+
+export interface deploymentArgs {
+  namespace: pulumi.Input<string>
+  domain: string
+  additionalDomain?: string
+  prometheusCreds: string
+}
+
+export class Ingress extends pulumi.ComponentResource {
+  constructor(name: string, args: deploymentArgs, opts?: pulumi.ComponentResourceOptions) {
+    super('prometheus', name, {}, opts)
+
+    const secretName = 'prometheus-cert-secret'
+
+    new k8s.core.v1.Secret(
+      `${name}-prometheus-auth-secret`,
+      {
+        metadata: {
+          name: `${name}-prometheus-auth-secret`,
+          namespace: args.namespace,
+        },
+        stringData: {
+          users: args.prometheusCreds,
+        },
+      },
+      { ...opts }
+    )
+
+    new k8s.apiextensions.CustomResource(
+      `${name}-prometheus-cert`,
+      {
+        apiVersion: 'cert-manager.io/v1',
+        kind: 'Certificate',
+        metadata: {
+          namespace: args.namespace,
+        },
+        spec: {
+          secretName: secretName,
+          duration: '2160h',
+          renewBefore: '360h',
+          isCA: false,
+          privateKey: {
+            algorithm: 'RSA',
+            encoding: 'PKCS1',
+            size: 2048,
+          },
+          dnsNames: [`prometheus.${args.domain}`],
+          issuerRef: {
+            name: 'lets-encrypt',
+            kind: 'ClusterIssuer',
+            group: 'cert-manager.io',
+          },
+        },
+      },
+      { ...opts }
+    )
+
+    const domains = args.additionalDomain
+      ? `Host(\`prometheus.${args.domain}\`) || Host(\`prometheus.${args.additionalDomain}\`)`
+      : `Host(\`prometheus.${args.domain}\`)`
+
+    new k8s.apiextensions.CustomResource(
+      `${name}-prometheus-ingress-auth`,
+      {
+        apiVersion: 'traefik.containo.us/v1alpha1',
+        kind: 'Middleware',
+        metadata: {
+          name: `${name}-prometheus-ingress-auth`,
+          namespace: args.namespace,
+        },
+        spec: {
+          basicAuth: {
+            secret: `${name}-prometheus-auth-secret`,
```

Comment:
you can assign the resource to a variable and access the metadata for resource names as an alternative to reduce possible name change discrepancy in the future `secret.metadata.name`. probably won't ever change but just an fyi for reference